build/utf8/valid_asm.go

```diff
@@ -0,0 +1,547 @@
+	JAE(LabelRef("test_first"))      //   more testing to do
+	ADDQ(Imm(1), i)                  // else: i++
+	JMP(LabelRef("start_utf8_loop")) //   continue
+
+	Label("test_first")
+	x := GP8()
+	MOVB(first.Idx(pi, 1), x)         // x = first[pi]
+	CMPB(x, Imm(xx))                  // if x == xx
+	JEQ(LabelRef("stdlib_ret_false")) //   return false (illegal started byte)
+
+	size := GP64()
+	MOVBQZX(x, size)     // size = x
+	ANDQ(Imm(0x7), size) // size &= 7
+	i2 := GP64()
+	MOVQ(i, i2)                      // i2 = i
+	ADDQ(size, i2)                   // i2 += size
+	CMPQ(i2, n)                      // if i2 > n
+	JA(LabelRef("stdlib_ret_false")) //  return false (short or invalid)
+
+	accept := GP16()
+	SHRB(Imm(4), x)                      // x = x >> 4
+	MOVW(acceptRanges.Idx(x, 2), accept) // accept = acceptRanges[x]
+	acceptLo := GP8()
+	MOVB(accept.As8(), acceptLo) // acceptLo = accept.lo
+	SHRW(Imm(8), accept)         // accept = accept.hi
+	// TODO: ^ this method to grab the fields seems odd
+
+	c := GP8()
+	MOVB(b.Offset(1), c)             // c = b[i+1]
+	CMPB(c, acceptLo)                // if c < accept.lo
+	JB(LabelRef("stdlib_ret_false")) //   return false
+	CMPB(accept.As8(), c)            // if accept.hi < c
+	JB(LabelRef("stdlib_ret_false")) //   return false
+
+	CMPQ(size, Imm(2))        // if size == 2
+	JEQ(LabelRef("inc_size")) //   -> inc_size
+
+	MOVB(b.Offset(2), c)             // c = b[i+2]
+	CMPB(c, U8(locb))                // if c < locb
+	JB(LabelRef("stdlib_ret_false")) //   return false
+	CMPB(c, U8(hicb))                // if hicb < c
+	JA(LabelRef("stdlib_ret_false")) //   return false
+
+	CMPQ(size, Imm(3))        // if size == 3
+	JEQ(LabelRef("inc_size")) //   -> inc_size
+
+	MOVB(b.Offset(3), c)             // c = b[i+3]
+	CMPB(c, Imm(locb))               // if c < locb
+	JB(LabelRef("stdlib_ret_false")) //   return false
+	CMPB(c, Imm(hicb))               // if hicb < c
+	JA(LabelRef("stdlib_ret_false")) //   return false
+
+	Label("inc_size")
+	ADDQ(size, i) // i += size
+
+	JMP(LabelRef("start_utf8_loop"))
+
+	Label("stdlib_ret_true")
+	MOVB(Imm(1), ret.Addr)
+	RET()
+	Label("stdlib_ret_false")
+	MOVB(Imm(0), ret.Addr)
+	RET()
+
+	Comment("End of stdlib implementation")
+}
+
+func incompleteMaskData() []byte {
+	// The incomplete mask is used on every block to flag the bytes that are
+	// incomplete if this is the last block (for example a byte that starts
+	// a 4 byte character only 3 bytes before the end).
+	any := byte(0xFF)
+	needs4 := byte(0b11110000) - 1
+	needs3 := byte(0b11100000) - 1
+	needs2 := byte(0b11000000) - 1
+	b := [32]byte{
+		any, any, any, any, any, any, any, any,
+		any, any, any, any, any, any, any, any,
+		any, any, any, any, any, any, any, any,
+		any, any, any, any, any, needs4, needs3, needs2,
+	}
+	return b[:]
+}
+
+func continuationMaskData(pattern byte) []byte {
+	// Pattern is something like 0b1110000 to accept all bytes of the form
```

Comment:
> Pattern is something like 0b1110000 to accept all bytes of the form

There are only 7 bits in this example, is this on purpose?

Reply:
definitely a typo